library/src/main/java/com/google/android/exoplayer2/text/eia608/Eia608Decoder.java

```diff
@@ -414,32 +501,201 @@ private void handleMiscCode(byte cc2) {
+    } else if (attribute >= 0x8 && attribute <= 0xF) {
+      // Attribute is an indent
+      if (cueIndent == DEFAULT_INDENT) {
+        // Only update the indent, if it's the default indent.
+        // This is not conform the spec, but otherwise indentations may be off
+        // because we don't create a new cue when we see the nextRowDown flag.
+        cueIndent = INDENT_MAP[attribute & 0x7];
+      }
+    }
+
+    // Parse the row bits
+    int row = cc1 & 0x7;
+    if (row >= 0x4) {
+      // Extended Preamble Code
+      row = row & 0x3;
+      switch (row) {
+        case 0x0:
+          // Row 14 or 15
+          cueLine = CUE_LINE_MAP[13];
+          break;
+        case 0x1:
+          // Row 5 or 6
+          cueLine = CUE_LINE_MAP[4];
+          break;
+        case 0x2:
+          // Row 7 or 8
+          cueLine = CUE_LINE_MAP[7];
+          break;
+        case 0x3:
+          // Row 9 or 10
+          cueLine = CUE_LINE_MAP[8];
+          break;
+      }
+    } else {
+      // Regular Preamble Code
+      switch (row) {
+        case 0x0:
+          // Row 11 (Default)
+          cueLine = CUE_LINE_MAP[10];
+          break;
+        case 0x1:
+          // Row 1 (Top)
+          cueLine = CUE_LINE_MAP[0];
+          break;
+        case 0x2:
+          // Row 4 (Top)
+          cueLine = CUE_LINE_MAP[3];
+          break;
+        case 0x3:
+          // Row 12 or 13 (Bottom)
+          cueLine = CUE_LINE_MAP[11];
+          break;
+      }
+    }
+  }
+
+  private void handleMidrowCode(byte cc1, byte cc2) {
+    boolean transparentOrUnderline = (cc2 & 0x1) != 0;
+    int attribute = cc2 >> 1 & 0xF;
+    if ((cc1 & 0x1) != 0) {
+      // Background Color
+      captionStyles.put(getSpanStartIndex(), new BackgroundColorSpan(transparentOrUnderline ?
+        COLOR_MAP[attribute] & TRANSPARENCY_MASK : COLOR_MAP[attribute]));
+    } else {
+      // Foreground color
+      captionStyles.put(getSpanStartIndex(), new ForegroundColorSpan(COLOR_MAP[attribute]));
+      if (transparentOrUnderline) {
+        // Text should be underlined
+        captionStyles.put(getSpanStartIndex(), new UnderlineSpan());
+      }
+    }
+  }
+
+  private void handleTabOffset(byte cc2) {
+    // Formula for tab offset handling adapted from:
+    // https://dvcs.w3.org/hg/text-tracks/raw-file/default/608toVTT/608toVTT.html#x1-preamble-address-code-pac
+    // We're ignoring any tab offsets that do not occur at the beginning of a new cue.
+    // This is not conform the spec, but works in most cases.
+    if (captionStringBuilder.length() == 0) {
+      switch (cc2) {
+        case CTRL_TAB_OFFSET_1:
+          tabOffset++;
+          break;
+        case CTRL_TAB_OFFSET_2:
+          tabOffset += 2;
+          break;
+        case CTRL_TAB_OFFSET_3:
+          tabOffset += 3;
+          break;
+      }
+    }
+  }
+
+  private int getSpanStartIndex() {
+    return captionStringBuilder.length() > 0 ? captionStringBuilder.length() - 1 : 0;
```

Comment:
Will this call length() twice or is the compiler clever enough to call it only once?